Mathlib/Geometry/Manifold/Riemannian/PathELength.lean

```diff
@@ -0,0 +1,365 @@
+import Mathlib.Geometry.Manifold.ContMDiff.NormedSpace
+import Mathlib.Geometry.Manifold.Instances.Icc
+import Mathlib.MeasureTheory.Constructions.UnitInterval
+import Mathlib.MeasureTheory.Function.JacobianOneDim
+
+/-! # Lengths of paths in manifolds
+
+Consider a manifold in which the tangent spaces have an enormed structure. Then one defines
+`pathELength γ a b` as the length of the path `γ : ℝ → M` between `a` and `b`, i.e., the integral
+of the norm of its derivative on `Icc a b`.
+
+We give several ways to write this quantity (as an integral over `Icc`, or `Ioo`, or the subtype
+`Icc`, using either `mfderiv` or `mfderivWithin`).
+
+We show that this notion is invariant under reparameterization by a monotone map, in
+`pathELength_comp_of_monotoneOn`.
+
+We define `riemannianEDist x y` as the infimum of the length of `C^1` paths between `x`
+and `y`. We show that this can be realized using `C^1` paths that are locally constant around the
+endpoints, in `exists_lt_locally_constant_of_riemannianEDist_lt`. Such paths can be glued while
+retaining the `C^1` property. We deduce that `riemannianEDist` satisfied the triangle inequality,
+in `riemannianEDist_triangle`.
+-/
+
+open Set MeasureTheory
+open scoped Manifold ENNReal ContDiff Topology
+
+noncomputable section
+
+variable
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E]
+  {H : Type*} [TopologicalSpace H] {I : ModelWithCorners ℝ E H} {n : WithTop ℕ∞}
+  {M : Type*} [TopologicalSpace M] [ChartedSpace H M]
+
+namespace Manifold
+
+variable [∀ (x : M), ENorm (TangentSpace I x)] {a b c a' b' : ℝ} {γ γ' : ℝ → M}
+
+variable (I) in
+/-- The length on `Icc a b` of a path into a manifold, where the path is defined on the whole real
+line.
+
+We use the whole real line to avoid subtype hell in API, but this is equivalent to
+considering functions on the manifold with boundary `Icc a b`, see
+`lintegral_norm_mfderiv_Icc_eq_pathELength_projIcc`.
+
+We use `mfderiv` instead of `mfderivWithin` in the definition as these coincide (apart from the two
+endpoints which have zero measure) and `mfderiv` is easier to manipulate. However, we give
+a lemma `pathELength_eq_integral_mfderivWithin_Icc` to rewrite with the `mfderivWithin` form. -/
+irreducible_def pathELength (γ : ℝ → M) (a b : ℝ) : ℝ≥0∞ :=
+  ∫⁻ t in Icc a b, ‖mfderiv 𝓘(ℝ) I γ t 1‖ₑ
+
+lemma pathELength_eq_lintegral_mfderiv_Icc :
+    pathELength I γ a b = ∫⁻ t in Icc a b, ‖mfderiv 𝓘(ℝ) I γ t 1‖ₑ := by simp [pathELength]
+
+lemma pathELength_eq_lintegral_mfderiv_Ioo :
+    pathELength I γ a b = ∫⁻ t in Ioo a b, ‖mfderiv 𝓘(ℝ) I γ t 1‖ₑ := by
+  rw [pathELength_eq_lintegral_mfderiv_Icc, restrict_Ioo_eq_restrict_Icc]
+
+lemma pathELength_eq_lintegral_mfderivWithin_Icc :
+    pathELength I γ a b = ∫⁻ t in Icc a b, ‖mfderivWithin 𝓘(ℝ) I γ (Icc a b) t 1‖ₑ := by
+  rw [pathELength_eq_lintegral_mfderiv_Icc, ← restrict_Ioo_eq_restrict_Icc]
+  apply setLIntegral_congr_fun measurableSet_Ioo (fun t ht ↦ ?_)
```

Comment:
Do you think adding a one-line comment about the (tiny) proof trick (or restricting to integration in Ioo, which is open, so mfderivWithin coincides with mfderiv) is worth it?